### ISSUE TITLE
Refused to evaluate a string as JavaScript because 'unsafe-eval' for Chrome Extension

### ISSUE THREAD

**Issue by pkuvegeta** [Polymer/polymer#1177]
I've tried using vulcanize with csp: true but it does not work.
Adding the following can work around this issue 
"content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'"

**Comment by vossad01** [comment]
I get this issue in a standard web application.

In Chrome, the offending line is (in polymer.js):
```js
var f = new Function('', 'return true;');
```

In Firefox, it is a very similar line, but it is in (webcomponents.js):
```js
var f = new Function("return true;");
```

I am using vulcanize with the CSP option.

There seem to be a lot of issues with trying to use Polymer with CSP [despite the documentation saying it works with CSP](https://www.polymer-project.org/resources/faq.html#csp).

**Comment by davidmaxwaterman** [comment]
What are the symptoms, specifically? IE, what does "does not work" mean?

**Comment by vossad01** [comment]
There are a number of CSP errors, between polymer and the polyfill.

Here is Chrome Canary (43.0.2314.0 canary (64-bit))...

with CSP:
![canary with](https://cloud.githubusercontent.com/assets/965439/6362775/4f45532e-bc56-11e4-84b8-1512992abb6f.JPG)

without CSP:
![canary without](https://cloud.githubusercontent.com/assets/965439/6362786/5d85f2c2-bc56-11e4-836b-1fc6b8d18ea4.JPG)

Granted most of the above is probably because of the 'unsafe-inline' not the 'inline-eval'. I would consider the fact it is throwing errors under CSP when CSP is documented as supported a bug in itself.

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <title>Polymer Demo</title>
  <script src="bower_components/webcomponentsjs/webcomponents.min.js"></script>
  <link rel="import" href="bower_components/paper-input/paper-input.html" />
</head>
<body>
  <label>Count:</label><paper-input></paper-input>
</body>
</html>
```

**Comment by davidmaxwaterman** [comment]
I thought the polyfill was only for non-chrome browsers...perhaps you don't need that at all?

Anyway, I didn't manage to get polymer elements working in a chrome extension (in the background.html page) - though I was only using core-ajax.

I have a vague memory that polymer does a quick 'test' attempt at something that doesn't comply with csp, and uses try-catch (or something) to adjust its behaviour accordingly, but it still does produce an error. It made me wonder if you were seeing that error - or perhaps it's not *just* that?

I'm also interested in seeing this work.

**Comment by vossad01** [comment]
While Chrome does not need the polyfill presentation issue happens on FF 36 too, which does need the polyfill.  Since the issues causing the screenshot I showed is different than the one originally reported, I will follow up elsewhere.

I have searched through the relevant code in polymer.js and webcomponents.js and it looks like this use of unsafe-eval is the test to see if eval is available and it will use eval as an optimization if it is.

I propose the resolution for this issue should be to add some sort of flag that can be set to indicate that CSP is in effect and Eval should not be used analogous to angular's [`ng-csp`](https://docs.angularjs.org/api/ng/directive/ngCsp) directive.
